### PR TITLE
Cmake overhaul update

### PR DESCRIPTION
**Pull request by crtrott**
Disable some warning, and fix CMAKE_CXX_EXTENSIONS warnings for CUDA builds.

### CODE REVIEW

**Comment by crtrott** [comment]
For some reason the "replace commas with semicolons" thing has to be done in cmake/kokkos_enable_options.cmake, which was called before it hits the cmake/kokkos_arch.cmake thing. Not sure why that replacement wasn't sticky ...

File: cmake/kokkos_cxx.cmake [hunk 1]
```diff
@@ -211,6 +188,37 @@ ELSE()
...
+# CMake recognizes the host compiler when calling nvcc_wrapper, this just
+# works.  Both NVCC and nvcc_wrapper only recognize '-std=c++11' which means
+# that we can only use host compilers for CUDA builds that use those flags.
+# It also means that extensions (gnu++11) can't be turned on for CUDA builds.
+IF(KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
+  IF(NOT DEFINED CMAKE_CXX_EXTENSIONS)
+    MESSAGE(FATAL_ERROR "NVCC doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
+  ELSEIF(CMAKE_CXX_EXTENSIONS)
+    MESSAGE(FATAL_ERROR "NVCC doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
+  ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_cxx.cmake]:
```CMake
if(NOT DEFINED CMAKE_CXX_EXTENSIONS OR CMAKE_CXX_EXTENSIONS)
```

File: cmake/kokkos_cxx.cmake [hunk 2]
```diff
@@ -211,6 +188,37 @@ ELSE()
...
+  # ENFORCE that the compiler can compile CUDA code.
+  IF(KOKKOS_CXX_COMPILER_ID STREQUAL Clang)
+    IF(KOKKOS_CXX_COMPILER_VERSION VERSION_LESS 4.0.0)
+      MESSAGE(FATAL_ERROR "Compiling CUDA code directly with Clang requires version 4.0.0 or higher.")
+    ENDIF()
+    IF(NOT DEFINED CMAKE_CXX_EXTENSIONS)
+      MESSAGE(FATAL_ERROR "Compiling CUDA code with clang doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
+    ELSEIF(CMAKE_CXX_EXTENSIONS)
+      MESSAGE(FATAL_ERROR "Compiling CUDA code with clang doesn't support C++ extensions.  Set -DCMAKE_CXX_EXTENSIONS=OFF")
+    ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_cxx.cmake]: Same comment here.

File: cmake/kokkos_cxx.cmake [hunk 3]
```diff
@@ -211,6 +188,37 @@ ELSE()
...
 ENDIF()
 
+# Enforce that extensions are turned off for nvcc_wrapper.
+# For compiling CUDA code using nvcc_wrapper, we will use the host compiler's
+# flags for turning on C++11.  Since for compiler ID and versioning purposes
+# CMake recognizes the host compiler when calling nvcc_wrapper, this just
+# works.  Both NVCC and nvcc_wrapper only recognize '-std=c++11' which means
+# that we can only use host compilers for CUDA builds that use those flags.
+# It also means that extensions (gnu++11) can't be turned on for CUDA builds.
+IF(KOKKOS_CXX_COMPILER_ID STREQUAL NVIDIA)
```
Comment (dalg24) [inline cmake/kokkos_cxx.cmake]: Any reason why that block wouldn't be within the `if(KOKKOS_ENABLE_CUDA)` block below?

File: cmake/kokkos_arch.cmake [hunk 4]
```diff
@@ -113,11 +113,11 @@ LIST(APPEND KOKKOS_ARCH_LIST
...
 IF (KOKKOS_ARCH MATCHES ",")
-  MESSAGE(WARNING "-- Detected a comma in: KOKKOS_ARCH=`${KOKKOS_ARCH}`")
-  MESSAGE("-- Although we prefer KOKKOS_ARCH to be semicolon-delimited, we do allow")
-  MESSAGE("-- comma-delimited values for compatibility with scripts (see github.com/trilinos/Trilinos/issues/2330)")
+#  MESSAGE(WARNING "-- Detected a comma in: KOKKOS_ARCH=`${KOKKOS_ARCH}`")
+#  MESSAGE("-- Although we prefer KOKKOS_ARCH to be semicolon-delimited, we do allow")
+#  MESSAGE("-- comma-delimited values for compatibility with scripts (see github.com/trilinos/Trilinos/issues/2330)")
   STRING(REPLACE "," ";" KOKKOS_ARCH "${KOKKOS_ARCH}")
-  MESSAGE("-- Commas were changed to semicolons, now KOKKOS_ARCH=`${KOKKOS_ARCH}`")
+#  MESSAGE("-- Commas were changed to semicolons, now KOKKOS_ARCH=`${KOKKOS_ARCH}`")
```
[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_arch.cmake]: Removed

File: cmake/kokkos_cxx.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_cxx.cmake]: Done

File: cmake/kokkos_cxx.cmake (hunk 2 above)
Comment (crtrott) [reply cmake/kokkos_cxx.cmake]: Done

File: cmake/kokkos_cxx.cmake (hunk 3 above)
Comment (crtrott) [reply cmake/kokkos_cxx.cmake]: In theory we could enable compilation with nvcc without CUDA being on (though it doesn't work right now) and NVCC doesn't know what gnu++11 is.

Comment (dalg24) [reply cmake/kokkos_cxx.cmake]: Is that something we want to support?

**Comment by crtrott** [comment]
I don't know about SEND_ERROR and we can change this to simply change the default to turn CMAKE_CXX_EXTENSIONS OFF if it is not defined. The reason for not making it a warning is that people don't read warnings ;-)

**Comment by jjwilke** [comment]
I agree it would be problematic to require explicitly turning off CMAKE_CXX_EXTENSIONS - particularly given how wonky things can get when reconfiguring from an existing cache.
I suspect most don't care whether they get extensions or not.

File: cmake/kokkos_cxx.cmake (hunk 3 above)
Comment (jjwilke) [reply cmake/kokkos_cxx.cmake]: This decision is above my pay grade.  I would personally throw either 1) throw an error in CUDA is not enable but nvcc is the compiler or 2) warn and enable CUDA.

File: cmake/kokkos_cxx.cmake (hunk 1 above)
Comment (jjwilke) [reply cmake/kokkos_cxx.cmake]: I would definitely change this to a warning for cmake extensions not being defined.

**Comment by dalg24** [comment]
Something to keep in mind is global properties like `CMAKE_CXX_EXTENSIONS` should be off limits.  We should be setting properties on targets.

Also here is another question worth asking ourselves:  do we want to enable compiler extensions at all?

**Pull request merged**